README.md
[![Build Status](https://secure.travis-ci.org/knyar/nginx-lua-prometheus.svg?branch=master)](http://travis-ci.org/knyar/nginx-lua-prometheus?branch=master)
[![Coverage Status](https://coveralls.io/repos/github/knyar/nginx-lua-prometheus/badge.svg?branch=master)](https://coveralls.io/github/knyar/nginx-lua-prometheus?branch=master)


# Prometheus metric library for Nginx

This is a Lua library that can be used with Nginx to keep track of metrics and
expose them on a separate web page to be pulled by
[Prometheus](https://prometheus.io).

## 使用 prometheus_wrapper.lua 快速接入

#### nginx.conf

```nginx
lua_shared_dict prometheus_metrics 10m;
server {
    # 添加到业务的 server 配置中
    # 如果有多个 server，可以传入 app 区分不同业务。比如，require("prometheus.wrapper"):log("new_app")
    log_by_lua_block {
        require("prometheus.wrapper"):log()
    }
}

# 单独配置
server {
    listen 8001;
    allow 10.0.0.0/8;
    allow 172.0.0.0/8;
    deny all;
    default_type  text/plain;
    location /metrics {
        content_by_lua_block {
            require("prometheus.wrapper"):metrics()
        }
    }
}
```

#### lua
添加到 init_by_lua_* 相关代码中，初始化相关配置
```lua
local ok, err = require("prometheus.wrapper"):init({
    app = "mbsug",
    idc = QIHOO_IDC,
    monitor_switch = {
        METRIC_COUNTER_RESPONSES = {"/s", "/status.html"},
        METRIC_COUNTER_SENT_BYTES = {"/s"},
        METRIC_COUNTER_REVD_BYTES = false,
        METRIC_HISTOGRAM_LATENCY = {"/s", "/jump"},
        METRIC_GAUGE_CONNECTS = true
    },
    log_method = {"GET", "POST"}, -- method 过滤
    buckets = {1,2,3,4,5,6,7,8,9,10,11,13,15,17,19,22,25,28,32,36,41,47,54,62,71,81,92,105,120,137,156,178,203,231,263,299,340,387,440,500} -- 桶距配置
    merge_path = "/metrics.php"
})

if not ok then
    ngx.log(ngx.ERR, "prometheus init error: ", err)
end
```

自定义 log，添加到 content_by_lua_* 相关代码中（如有需要）
```lua
local wrapper = require("prometheus.wrapper")
-- histogram
wrapper:latencyLog(2.23, "/get", "searcher", "GET") -- 延迟
-- counter
wrapper:qpsCounterLog(1, "/get", "searcher", "GET", 200) -- QPS
wrapper:sendBytesCounterLog(1024, "/get", "searcher", "GET", 200) -- 流量 out
wrapper:receiveBytesCounterLog(2048, "/get", "searcher", "GET", 200) -- 流量 in
wrapper:exceptionLog(1, "mysql_connect_err", "self") -- 异常
-- gauge
wrapper:gaugeLog("alive", "searcher") -- 状态
```

## Quick start guide

You would need to install nginx package with lua support (`nginx-extras` on
Debian) and make `prometheus.lua` available in your `LUA_PATH` (or just point
`lua_package_path` to a directory with this git repo).

To track request latency broken down by HTTP host and request count broken
down by host and status, add the following to `nginx.conf`:

```
lua_shared_dict prometheus_metrics 10M;
lua_package_path "/path/to/nginx-lua-prometheus/?.lua";
init_by_lua '
  prometheus = require("prometheus").init("prometheus_metrics")
  metric_requests = prometheus:counter(
    "nginx_http_requests_total", "Number of HTTP requests", {"host", "status"})
  metric_latency = prometheus:histogram(
    "nginx_http_request_duration_seconds", "HTTP request latency", {"host"})
  metric_connections = prometheus:gauge(
    "nginx_http_connections", "Number of HTTP connections", {"state"})
';
log_by_lua '
  local host = ngx.var.host:gsub("^www.", "")
  metric_requests:inc(1, {host, ngx.var.status})
  metric_latency:observe(ngx.now() - ngx.req.start_time(), {host})
';
```

This:
* configures a shared dictionary for your metrics called `prometheus_metrics`
  with a 10MB size limit;
* registers a counter called `nginx_http_requests_total` with two labels:
  `host` and `status`;
* registers a histogram called `nginx_http_request_duration_seconds` with one
  label `host`;
* registers a gauge called `nginx_http_connections` with one label `state`;
* on each HTTP request measures its latency, recording it in the histogram and
  increments the counter, setting current HTTP host as `host` label and
  HTTP status code as `status` label.

Last step is to configure a separate server that will expose the metrics.
Please make sure to only make it reachable from your Prometheus server:

```
server {
  listen 9145;
  allow 192.168.0.0/16;
  deny all;
  location /metrics {
    content_by_lua '
      metric_connections:set(ngx.var.connections_reading, {"reading"})
      metric_connections:set(ngx.var.connections_waiting, {"waiting"})
      metric_connections:set(ngx.var.connections_writing, {"writing"})
      prometheus:collect()
    ';
  }
}
```

Metrics will be available at `http://your.nginx:9145/metrics`. Note that the
gauge metric in this example contains values obtained from nginx global state,
so they get set immediately before metrics are returned to the client.

**Note**: using HTTP host as a metric label value on servers that have many
virtual hosts has potential performance implications. Please read the caveats
section below for more information.

## API reference

### init()

**syntax:** require("prometheus").init(*dict_name*, [*prefix*])

Initializes the module. This should be called once from the
[init_by_lua](https://github.com/openresty/lua-nginx-module#init_by_lua)
section in nginx configuration.

* `dict_name` is the name of the nginx shared dictionary which will be used to
  store all metrics. Defaults to `prometheus_metrics` if not specified.
* `prefix` is an optional string which will be prepended to metric names on output


Returns a `prometheus` object that should be used to register metrics.

Example:
```
init_by_lua '
  prometheus = require("prometheus").init("prometheus_metrics")
';
```

### prometheus:counter()

**syntax:** prometheus:counter(*name*, *description*, *label_names*)

Registers a counter. Should be called once from the
[init_by_lua](https://github.com/openresty/lua-nginx-module#init_by_lua)
section.

* `name` is the name of the metric.
* `description` is the text description that will be presented to Prometheus
  along with the metric. Optional (pass `nil` if you still need to define
  label names).
* `label_names` is an array of label names for the metric. Optional.

[Naming section](https://prometheus.io/docs/practices/naming/) of Prometheus
documentation provides good guidelines on choosing metric and label names.

Returns a `counter` object that can later be incremented.

Example:
```
init_by_lua '
  prometheus = require("prometheus").init("prometheus_metrics")
  metric_bytes = prometheus:counter(
    "nginx_http_request_size_bytes", "Total size of incoming requests")
  metric_requests = prometheus:counter(
    "nginx_http_requests_total", "Number of HTTP requests", {"host", "status"})
';
```

### prometheus:gauge()

**syntax:** prometheus:gauge(*name*, *description*, *label_names*)

Registers a gauge. Should be called once from the
[init_by_lua](https://github.com/openresty/lua-nginx-module#init_by_lua)
section.

* `name` is the name of the metric.
* `description` is the text description that will be presented to Prometheus
  along with the metric. Optional (pass `nil` if you still need to define
  label names).
* `label_names` is an array of label names for the metric. Optional.

Returns a `gauge` object that can later be set.

Example:
```
init_by_lua '
  prometheus = require("prometheus").init("prometheus_metrics")
  metric_connections = prometheus:gauge(
    "nginx_http_connections", "Number of HTTP connections", {"state"})
';
```

### prometheus:histogram()

**syntax:** prometheus:histogram(*name*, *description*, *label_names*,
  *buckets*)

Registers a histogram. Should be called once from the
[init_by_lua](https://github.com/openresty/lua-nginx-module#init_by_lua)
section.

* `name` is the name of the metric.
* `description` is the text description. Optional.
* `label_names` is an array of label names for the metric. Optional.
* `buckets` is an array of numbers defining bucket boundaries. Optional,
  defaults to 20 latency buckets covering a range from 5ms to 10s (in seconds).

Returns a `histogram` object that can later be used to record samples.

Example:
```
init_by_lua '
  prometheus = require("prometheus").init("prometheus_metrics")
  metric_latency = prometheus:histogram(
    "nginx_http_request_duration_seconds", "HTTP request latency", {"host"})
  metric_response_sizes = prometheus:histogram(
    "nginx_http_response_size_bytes", "Size of HTTP responses", nil,
    {10,100,1000,10000,100000,1000000})
';
```

### prometheus:collect()

**syntax:** prometheus:collect()

Presents all metrics in a text format compatible with Prometheus. This should be
called in
[content_by_lua](https://github.com/openresty/lua-nginx-module#content_by_lua)
to expose the metrics on a separate HTTP page.

Example:
```
location /metrics {
  content_by_lua 'prometheus:collect()';
  allow 192.168.0.0/16;
  deny all;
}
```

### counter:inc()

**syntax:** counter:inc(*value*, *label_values*)

Increments a previously registered counter. This is usually called from
[log_by_lua](https://github.com/openresty/lua-nginx-module#log_by_lua)
globally or per server/location.

* `value` is a value that should be added to the counter. Defaults to 1.
* `label_values` is an array of label values.

The number of label values should match the number of label names defined when
the counter was registered using `prometheus:counter()`. No label values should
be provided for counters with no labels. Non-printable characters will be
stripped from label values.

Example:
```
log_by_lua '
  metric_bytes:inc(tonumber(ngx.var.request_length))
  metric_requests:inc(1, {ngx.var.host, ngx.var.status})
';
```

### gauge:set()

**syntax:** gauge:set(*value*, *label_values*)

Sets the current value of a previously registered gauge. This could be called
from [log_by_lua](https://github.com/openresty/lua-nginx-module#log_by_lua)
globally or per server/location to modify a gauge on each request, or from
[content_by_lua](https://github.com/openresty/lua-nginx-module#content_by_lua)
just before `prometheus::collect()` to return a real-time value.

* `value` is a value that the gauge should be set to. Required.
* `label_values` is an array of label values.

### histogram:observe()

**syntax:** histogram:observe(*value*, *label_values*)

Records a value in a previously registered histogram. Usually called from
[log_by_lua](https://github.com/openresty/lua-nginx-module#log_by_lua)
globally or per server/location.

* `value` is a value that should be recorded. Required.
* `label_values` is an array of label values.

Example:
```
log_by_lua '
  metric_latency:observe(ngx.now() - ngx.req.start_time(), {ngx.var.host})
  metric_response_sizes:observe(tonumber(ngx.var.bytes_sent))
';
```

### Built-in metrics

The module increments the `nginx_metric_errors_total` metric if it encounters
an error (for example, when `lua_shared_dict` becomes full). You might want
to configure an alert on that metric.

## Caveats

Please keep in mind that all metrics stored by this library are kept in a
single shared dictionary (`lua_shared_dict`). While exposing metrics the module
has to list all dictionary keys, which has serious performance implications for
dictionaries with large number of keys (in this case this means large number
of metrics OR metrics with high label cardinality). Listing the keys has to
lock the dictionary, which blocks all threads that try to access it (i.e.
potentially all nginx worker threads).

There is no elegant solution to this issue (besides keeping metrics in a
separate storage system external to nginx), so for latency-critical servers you
might want to keep the number of metrics (and distinct metric label values) to
a minimum.

## Development

### Install dependencies for testing

- `luarocks install luacheck`
- `luarocks install luaunit`

### Run tests

- `luacheck --globals ngx -- prometheus.lua`
- `lua prometheus_test.lua`

## Credits

- Created and maintained by Anton Tolchanov (@knyar)
- Metrix prefix support contributed by david birdsong (@davidbirdsong)
- Gauge support contributed by Cosmo Petrich (@cosmopetrich)

## License

Licensed under MIT license.


CHANGELOG.md
## 1.1.4 / 2017-08-31

  * 修复多参数 uri 匹配问题


## 1.1.3 / 2017-06-28

  * log 阶段添加 app 参数，支持多服务区分
  * 自定义 log 添加 app 参数
  * 添加 debug 配置，方便开发


## 1.1.2 / 2017-06-27

  * Bug Fixes


## 1.1.1 / 2017-06-26

  * uri 配置支持带参数
  * 优化部分代码逻辑
  * 新增异常统计 counter


## 1.1.0 / 2017-06-24

  * 原始 SDK 更新


## 1.0.0 / 2017-06-24

  * 添加自定义 Log 入口
  * 去除全局变量的使用
  * 完善文档


## 0.0.3 / 2017-06-16

  * 优化部分代码逻辑
  * 合并统计项目去除注释，避免数据收集报错


## 0.0.2 / 2017-06-09

  * 添加 `metric_traffic_in` 监控项
  * 添加配置开关，自定义开启关闭统计项
  * 添加合并下游统计项功能


## 0.0.1 / 2017-06-07

  * Wrapper SDK 初始化，各流程统一封装，方便接入


prometheus_wrapper.lua
-- Copyright (C) by Jiang Yang (jiangyang-pd@360.cn)

local _M = { _VERSION = "0.0.1" }

_M.CONF = {
    initted = false,
    app = "default",
    idc = "",
    counter_path = {},
    histogram_path = {},
    log_method = {},
    buckets = {10,11,12,14,16,18,20,23,26,29,33,37,42,47,53,60,67,75,84,95,107,120,135,152,171,192,216,243,273,307,345,388,437,492,554,623,701,789,888,1000}
}

local function inTable(needle, table_name)
    if type(needle) ~= "string" or type(table_name) ~= "table" then
        return false
    end
    for _, v in ipairs(table_name) do
        if v == needle then
            return true
        end
    end
    return false
end

function _M:init(user_config)
    for k, v in pairs(user_config) do
        if k == "app" then
            if type(v) ~= "string" then
                return nil, '"app" must be a string'
            end
            self.CONF.app = v
        elseif k == "idc" then
            if type(v) ~= "string" then
                return nil, '"idc" must be a string'
            end
            self.CONF.idc = v
        elseif k == "counter_path" then
            if type(v) ~= "table" then
                return nil, '"counter_path" must be a table'
            end
            self.CONF.counter_path = v
        elseif k == "histogram_path" then
            if type(v) ~= "table" then
                return nil, '"histogram_path" must be a table'
            end
            self.CONF.histogram_path = v
        elseif k == "log_method" then
            if type(v) ~= "table" then
                return nil, '"log_method" must be a table'
            end
            self.CONF.log_method = v
        elseif k == "buckets" then
            if type(v) ~= "table" then
                return nil, '"buckets" must be a table'
            end
            self.CONF.buckets = v
        end
    end

    if true then
        self.CONF.initted = true
    end

    local config = ngx.shared.prometheus_metrics
    config:flush_all()

    prometheus = require("prometheus.prometheus").init("prometheus_metrics")

    -- QPS
    metric_requests = prometheus:counter(
        "module_responses", 
        "[" .. self.CONF.idc .. "] number of /path", 
        {"app", "api", "module", "method", "code"}
    )

    -- 流量
    metric_traffic = prometheus:counter(
        "module_sent_bytes",
        "[" .. self.CONF.idc .. "] traffic of /path",
        {"app", "api", "module", "method", "code"}
    )

    -- 延迟
    metric_latency = prometheus:histogram(
        "response_duration_milliseconds",
        "[" .. self.CONF.idc .. "] http request latency", 
        {"app", "api", "module", "method"},
        self.CONF.buckets
    )

    -- 状态
    metric_connections = prometheus:gauge(
        "module_connections",
        "[" .. self.CONF.idc .. "] number of HTTP connections", 
        {"app", "state"}
    )

    return self.CONF.initted
end

function _M:log()
    if not self.CONF.initted then
        return nil, "init first.."
    end

    local path = ""
    local method = ngx.var.request_method or ""
    local request_uri = ngx.var.request_uri or ""
    local status = ngx.var.status or ""

    local pathInCounter = false
    local pathInHistogram = false

    if not request_uri or not method then
        return nil, "empty request_uri|method"
    end

    local st, sp = string.find(request_uri, "?")
    if st == nil then
        path = request_uri
    else
        path = string.sub(request_uri, 1, st-1)
    end

    path = string.lower(path)
    pathInCounter = inTable(path, self.CONF.counter_path)
    pathInHistogram = inTable(path, self.CONF.histogram_path)

    if (pathInCounter or pathInHistogram) and inTable(method, self.CONF.log_method) then
        if pathInCounter then
            metric_requests:inc(1, {self.CONF.app, path, "self", method, status})
            metric_traffic:inc(tonumber(ngx.var.bytes_sent), {self.CONF.app, path, "self", method, status})
        end

        if pathInHistogram then
            local tm = (ngx.now() - ngx.req.start_time()) * 1000
            metric_latency:observe(tm, {self.CONF.app, path, "self", method})
        end
    end

    return true
end

function _M:metrics()
    local ip = ngx.var.remote_addr or ""
    local st, sp = string.find(ip, ".", 1, true)
    local sub_ip = ip
    if st == nil then
        sub_ip = ip
    else
        sub_ip = string.sub(ip, 1, st-1)
    end

    if sub_ip ~= '10' and sub_ip ~= '172' then
        ngx.exit(ngx.HTTP_FORBIDDEN)
    end

    if not self.CONF.initted then
        ngx.say("init first..")
        ngx.exit(ngx.HTTP_OK)
    end

    if ngx.var.connections_reading and ngx.var.connections_waiting and ngx.var.connections_writing then
        metric_connections:set(ngx.var.connections_reading, {self.CONF.app, "reading"})
        metric_connections:set(ngx.var.connections_waiting, {self.CONF.app, "waiting"})
        metric_connections:set(ngx.var.connections_writing, {self.CONF.app, "writing"})
    end
    
    prometheus:collect()
end

return _M


wrapper.lua
-- Copyright (C) by Jiang Yang (jiangyang-pd@360.cn)

local _M = { _VERSION = "1.1.3" }

local find = string.find
local sub = string.sub

_M.CONF = {
    initted = false,
    app = "default",
    idc = "",
    monitor_switch = {
        METRIC_COUNTER_RESPONSES = {},
        METRIC_COUNTER_SENT_BYTES = {},
        METRIC_COUNTER_REVD_BYTES = {},
        METRIC_HISTOGRAM_LATENCY = {},
        METRIC_COUNTER_EXCEPTION = true,
        METRIC_GAUGE_CONNECTS = true,
    },
    log_method = {},
    buckets = {},
    merge_path = false,
    debug = false
}


local function inTable(needle, table_name)
    if type(needle) ~= "string" or type(table_name) ~= "table" then
        return false
    end
    for _, v in ipairs(table_name) do
        if v == needle then
            return true
        end
    end
    return false
end


local function empty(var)
    if type(var) == "table" then
        return next(var) == nil
    end
    return var == nil or var == '' or not var
end


local function explode(separator, str)
    if (separator == '') then return false end
    local pos, t = 0, {}
    -- for each divider found
    for st,sp in function() return find(str, separator, pos, true) end do
        table.insert(t, sub(str, pos, st-1)) -- Attach chars left of current divider
        pos = sp + 1 -- Jump past current divider
    end
    table.insert(t, sub(str, pos)) -- Attach chars right of last divider
    return t
end


function _M:init(user_config)
    for k, v in pairs(user_config) do
        if k == "app" then
            if type(v) ~= "string" then
                return nil, '"app" must be a string'
            end
            self.CONF.app = v
        elseif k == "idc" then
            if type(v) ~= "string" then
                return nil, '"idc" must be a string'
            end
            self.CONF.idc = v
        elseif k == "log_method" then
            if type(v) ~= "table" then
                return nil, '"log_method" must be a table'
            end
            self.CONF.log_method = v
        elseif k == "buckets" then
            if type(v) ~= "table" then
                return nil, '"buckets" must be a table'
            end
            self.CONF.buckets = v
        elseif k == "monitor_switch" then
            if type(v) ~= "table" then
                return nil, '"monitor_switch" must be a table'
            end
            for i, j in pairs(v) do
                if type(self.CONF.monitor_switch[i]) == "table" then
                    self.CONF.monitor_switch[i] = j
                end
            end
        elseif k == "merge_path" then
            if type(v) ~= "string" then
                return nil, '"merge_path" must be a string'
            end
            self.CONF.merge_path = v
        elseif k == "debug" then
            if type(v) ~= "boolean" then
                return nil, '"debug" must be a boolean'
            end
            self.CONF.debug = v
        end
    end

    if self.CONF.debug == false then
        local config = ngx.shared.prometheus_metrics
        config:flush_all()
    end

    local prometheus = require("prometheus.prometheus").init("prometheus_metrics")

    -- QPS
    if not empty(self.CONF.monitor_switch.METRIC_COUNTER_RESPONSES) then
        self:parseLogUri("METRIC_COUNTER_RESPONSES")
        self.metric_requests = prometheus:counter(
            "module_responses",
            "[" .. self.CONF.idc .. "] number of /path",
            {"app", "api", "module", "method", "code"}
        )
    end

    -- 流量 out
    if not empty(self.CONF.monitor_switch.METRIC_COUNTER_SENT_BYTES) then
        self:parseLogUri("METRIC_COUNTER_SENT_BYTES")
        self.metric_traffic_out = prometheus:counter(
            "module_sent_bytes",
            "[" .. self.CONF.idc .. "] traffic out of /path",
            {"app", "api", "module", "method", "code"}
        )
    end

    -- 流量 in
    if not empty(self.CONF.monitor_switch.METRIC_COUNTER_REVD_BYTES) then
        self:parseLogUri("METRIC_COUNTER_REVD_BYTES")
        self.metric_traffic_in = prometheus:counter(
            "module_rcvd_bytes",
            "[" .. self.CONF.idc .. "] traffic in of /path",
            {"app", "api", "module", "method", "code"}
        )
    end

    -- 程序异常
    if not empty(self.CONF.monitor_switch.METRIC_COUNTER_EXCEPTION) then
        self.metric_exceptions = prometheus:counter(
            "module_exceptions",
            "[" .. self.CONF.idc .. "] exceptions",
            {"app", "exception", "module"}
        )
    end

    -- 延迟
    if not empty(self.CONF.monitor_switch.METRIC_HISTOGRAM_LATENCY) then
        self:parseLogUri("METRIC_HISTOGRAM_LATENCY")
        self.metric_latency = prometheus:histogram(
            "response_duration_milliseconds",
            "[" .. self.CONF.idc .. "] http request latency",
            {"app", "api", "module", "method"},
            self.CONF.buckets
        )
    end

    -- 状态
    if not empty(self.CONF.monitor_switch.METRIC_GAUGE_CONNECTS) then
        self.metric_connections = prometheus:gauge(
            "module_connections",
            "[" .. self.CONF.idc .. "] state",
            {"app", "state"}
        )
    end

    if true then
        self.CONF.initted = true
        self.prometheus = prometheus
    end

    return self.CONF.initted
end


function _M:log(app)
    if not self.CONF.initted then
        return nil, "init first.."
    end

    local uri
    local app = app or self.CONF.app
    local method = ngx.var.request_method or ""
    local request_uri = ngx.var.request_uri or ""
    local status = ngx.var.status or ""

    if not request_uri or not method then
        return nil, "empty request_uri|method"
    end
    request_uri = string.lower(request_uri)

    if inTable(method, self.CONF.log_method) then
        uri = self:isLogUri(request_uri, "METRIC_COUNTER_RESPONSES")
        if self.metric_requests and uri then
            self.metric_requests:inc(1, {app, uri, "self", method, status})
        end

        uri = self:isLogUri(request_uri, "METRIC_COUNTER_SENT_BYTES")
        if self.metric_traffic_out and uri then
            self.metric_traffic_out:inc(tonumber(ngx.var.bytes_sent), {app, uri, "self", method, status})
        end

        uri = self:isLogUri(request_uri, "METRIC_COUNTER_REVD_BYTES")
        if self.metric_traffic_in and uri then
            self.metric_traffic_in:inc(tonumber(ngx.var.request_length), {app, uri, "self", method, status})
        end

        uri = self:isLogUri(request_uri, "METRIC_HISTOGRAM_LATENCY")
        if self.metric_latency and uri then
            local tm = (ngx.now() - ngx.req.start_time()) * 1000
            self.metric_latency:observe(tm, {app, uri, "self", method})
        end
    end

    return true
end


function _M:latencyLog(time, api, module_name, method, app)
    if not self.metric_latency or not self.CONF.initted then
        return false
    end
    app = app or self.CONF.app
    method = method or "GET"
    self.metric_latency:observe(time, {app, api, module_name, method})
    return true
end


function _M:counterLog(counter_ins, value, api, module_name, method, code, app)
    if not counter_ins or not self.CONF.initted then
        return false
    end
    method = method or "GET"
    code = code or 200
    module_name = module_name or "self"
    app = app or self.CONF.app
    counter_ins:inc(tonumber(value), {app, api, module_name, method, code})
    return true
end


function _M:qpsCounterLog(times, api, module_name, method, code, app)
    return self:counterLog(self.metric_requests, times, api, module_name, method, code, app)
end


function _M:sendBytesCounterLog(bytes, api, module_name, method, code, app)
    return self:counterLog(self.metric_traffic_out, bytes, api, module_name, method, code, app)
end


function _M:receiveBytesCounterLog(bytes, api, module_name, method, code, app)
    return self:counterLog(self.metric_traffic_in, bytes, api, module_name, method, code, app)
end


function _M:exceptionLog(times, exception, module_name, app)
    if not self.metric_exceptions or not self.CONF.initted then
        return false
    end
    module_name = module_name or "self"
    app = app or self.CONF.app
    self.metric_exceptions:inc(tonumber(times), {app, exception, module_name})
    return true
end


function _M:gaugeLog(value, state, app)
    if not self.metric_connections or not self.CONF.initted then
        return false
    end
    app = app or self.CONF.app
    self.metric_connections:set(value, {app, state})
    return true
end


function _M:getPrometheus()
    if not self.CONF.initted then
        return nil, "init first.."
    end
    return self.prometheus
end


function _M:parseLogUri(monitor_key)
    local res = {}
    local _t = type(self.CONF.monitor_switch[monitor_key])
    if _t == "table" then
        for _, uri in ipairs(self.CONF.monitor_switch[monitor_key]) do
            -- /idxdata/get?type=obx&name=test => /idxdata/get, {"type=obx", name=test}
            local path, params = self:parseUri(uri)
            local uriConf = {
                uri = uri,
                path = path,
                params = params
            }
            table.insert(res, uriConf)
        end
    elseif _t == "boolean" then
        res = self.CONF.monitor_switch[monitor_key]
    else
        res = false
    end
    self.CONF.monitor_switch[monitor_key] = res
end


function _M:isLogUri(request_uri, monitor_key)
    if type(self.CONF.monitor_switch[monitor_key]) ~= "table" then
        return false
    end
    local request_path, request_params = self:parseUri(request_uri)
    for _, uriConf in ipairs(self.CONF.monitor_switch[monitor_key]) do
        if uriConf["path"] == request_path then
            if empty(uriConf["params"]) then
                return uriConf["uri"]
            else
                for _, param in ipairs(uriConf["params"]) do
                    if not inTable(param, request_params) then
                        return false
                    end
                end
                return uriConf["uri"]
            end
        end
    end
    return false
end


function _M:parseUri(uri)
    local path = ""
    local params = {}
    local st, _ = find(uri, "?")
    if st == nil then
        path = uri
    else
        path = sub(uri, 1, st-1)
        local param = sub(uri, st+1)
        if param and param ~= "" then
            params = explode("&", param)
        end
    end
    return path, params
end


function _M:metrics()
    local ip = ngx.var.remote_addr or ""
    local st, _ = find(ip, ".", 1, true)
    local sub_ip = ip
    if st == nil then
        sub_ip = ip
    else
        sub_ip = sub(ip, 1, st-1)
    end

    if sub_ip ~= '10' and sub_ip ~= '172' then
        ngx.exit(ngx.HTTP_FORBIDDEN)
    end

    if not self.CONF.initted then
        ngx.say("init first..")
        ngx.exit(ngx.HTTP_OK)
    end

    if self.metric_connections and ngx.var.connections_reading and ngx.var.connections_waiting and ngx.var.connections_writing then
        self.metric_connections:set(ngx.var.connections_reading, {self.CONF.app, "reading"})
        self.metric_connections:set(ngx.var.connections_waiting, {self.CONF.app, "waiting"})
        self.metric_connections:set(ngx.var.connections_writing, {self.CONF.app, "writing"})
    end

    self.prometheus:collect()

    -- 合并下游自定义统计项, merge_path 需跟 metrics 在同一个server下
    if self.CONF.merge_path and type(self.CONF.merge_path) == "string" then
        local res = ngx.location.capture(self.CONF.merge_path)
        if res and res.status == 200 and type(res.body) == "string" and res.body then
            local newstr, _, err = ngx.re.gsub(res.body, "# (HELP|TYPE).*\n", "", "i")
            if newstr then
                ngx.say(newstr)
            else
                ngx.log(ngx.ERR, "error: ", err)
            end
        end
    end
end


return _M
